### PR TITLE
Build and Sign on VSTS

### PR DESCRIPTION
**Pull request by shhsu**

### CODE REVIEW

File: azure-pipelines/powershell/Copy-SignedBits.ps1 [hunk 1]
```diff
@@ -0,0 +1,51 @@
+[cmdletbinding()]
```
Comment (shhsu) [inline azure-pipelines/powershell/Copy-SignedBits.ps1]: Talked to @bangbingsyb , we want to investigate how to remove this script and just have msbuild publish things in the right place

File: azure-pipelines/verify.signing/no_strongname.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+*.dll, strong name signing not needed
```
Comment (bangbingsyb) [inline azure-pipelines/verify.signing/no_strongname.txt]: Since you don't want to do any strong name validation, instead of having a whitelist to exclude everything, you can configure the task as：

ExcludeSNVerify: true

See the [doc](https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki?wikiVersion=GBwikiMaster&pagePath=%2FEngineering%20System%20%26%20Tools%2FMicroBuild%2FEnabling%20Signing%20in%20Your%20Projects%2FCodesign%20Verification&pageId=653)

File: build/Microsoft.IIS.Administration.props [hunk 3]
```diff
@@ -7,6 +7,8 @@
     <PublishRoot>$(SolutionDir)dist</PublishRoot>
     <AppPublishRoot>$(SolutionDir)dist\Microsoft.IIS.Administration</AppPublishRoot>
     <PluginsPublishRoot>$(SolutionDir)dist\Microsoft.IIS.Administration\plugins</PluginsPublishRoot>
+    <SignAssembly>true</SignAssembly>
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
```
Comment (bangbingsyb) [inline build/Microsoft.IIS.Administration.props]: Why is the condition " "$(PublishDir) == '' " removed for CopyLocalLockFileAssemblies?

File: build/sign.props [hunk 4]
```diff
@@ -10,4 +10,14 @@
       <Authenticode>$(SigningIdentity)</Authenticode>
     </FilesToSign>
   </ItemGroup>
-</Project>
\ No newline at end of file
+
+  <!-- Include this segment to debug signing -->
```
Comment (bangbingsyb) [inline build/sign.props]: Should we remove this section before check in?

File: azure-pipelines/verify.signing/no_strongname.txt (hunk 2 above)
Comment (shhsu) [reply azure-pipelines/verify.signing/no_strongname.txt]: done

File: build/Microsoft.IIS.Administration.props (hunk 3 above)
Comment (shhsu) [reply build/Microsoft.IIS.Administration.props]: It's a long story:

We used to have a "packager" project that includes all plugins as assembly and that project would evaluate CopyLocalLockFileAssemblies to true in publish. Effective we were still doing CopyLocalLockFileAssemblies. That statement were there just to prevent the redundant operations.

That packager was causing a bunch of weird runtime behavior and essentially was just there the facilitate the .\publish.ps1 script I mentioned. The whole process is really wonky and now that I have removed packager and have each plugin build on its own, we no longer have that condition.

In short: We were effectively always doing it

File: build/sign.props (hunk 4 above)
Comment (shhsu) [reply build/sign.props]: I think it should be checked in. For future debugging convenience.

However, including this un-commented would cause "dotnet restore" to fail on azure pipeline. For some reasons I didn't bother to investigate it has issues pulling the package from microbuild source. (maybe credentials)
Also any developers external to microsoft won't be able to build if it is un-commented.

File: build/Microsoft.IIS.Administration.props (hunk 3 above)
Comment (shhsu) [reply build/Microsoft.IIS.Administration.props]: The condition `<PropertyGroup Condition="$(PublishDir) == ''">` has no effect, `CopyLocalLockFileAssemblies` is true on publish anyway...

File: build/sign.props (hunk 4 above)
Comment (bangbingsyb) [reply build/sign.props]: That's exactly the reason I suggest remove it and not check this part in - "Also any developers external to microsoft won't be able to build if it is un-commented."

For an open-source project, we shouldn't include random things for internal debugging usage - it will simply confuse the readers. You can simply keep this section in your onenote - how many times do we really need to debug microbuild signing?

File: build/Microsoft.IIS.Administration.props (hunk 3 above)
Comment (bangbingsyb) [reply build/Microsoft.IIS.Administration.props]: Then how about dotnet build - is CopyLocalLockFileAssemblies always true as well?

Comment (shhsu) [reply build/Microsoft.IIS.Administration.props]: From our existing `plugins.props` we would always set `CopyLocalLockFileAssemblies` on true. This line basically continues the same behavior

File: build/sign.props (hunk 4 above)
Comment (shhsu) [reply build/sign.props]: ok, removing in next revision

File: azure-pipelines/scripts/Copy-SignedBits.ps1 [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    if (!$publishDir) {
+        $publishDir = Join-Path $projectRoot "dist"
+    }
+}
+
+Push-Location $buildDir
+Write-Verbose "Locate build directory $buildDir"
+try {
+    foreach ($bitPath in Get-ChildItem -Recurse -Filter *.dll | Resolve-Path -Relative) {
```
Comment (bangbingsyb) [inline azure-pipelines/scripts/Copy-SignedBits.ps1]: Do you have any exe or dlls only?

File: src/Microsoft.IIS.Administration/Microsoft.IIS.Administration.csproj [hunk 6]
```diff
@@ -106,6 +106,10 @@
     <Exec Command="powershell -NonInteractive -executionpolicy Unrestricted -command &quot;&amp;'..\..\scripts\build\Copy-Configs.ps1' '$(MSBuildProjectDirectory)'&quot;" />
   </Target>
 
+  <Target Name="CleanBuildDir" Condition="$(CleanBuildDir) != ''" AfterTargets="Build">
+    <Exec Command="powershell -NonInteractive -executionpolicy Unrestricted -command &quot;&amp;..\..\scripts\build\Clean-BuildDir.ps1&quot;" />
```
Comment (bangbingsyb) [inline src/Microsoft.IIS.Administration/Microsoft.IIS.Administration.csproj]: Nit: recommend to use absolute path with $(SolutionDir) instead of a relative path for readability and in case the project file is moved to another location.

**Review by bangbingsyb — APPROVED** [review]
